### PR TITLE
Fixing multiple tokenizations in same experiment directory bug

### PR DESCRIPTION
**Pull request by pruksmhc**
https://github.com/nyu-mll/jiant/issues/862

### CODE REVIEW

File: jiant/preprocess.py [hunk 1]
```diff
@@ -282,9 +284,9 @@ def build_tasks(args):
         setattr(task, "_classifier_name", task_classifier if task_classifier else task.name)
 
     tokenizer_names = {task.name: task.tokenizer_name for task in tasks}
-    assert len(set(tokenizer_names.values())) == 1, (
-        f"Error: mixing tasks with different tokenizers!" " Tokenizations: {tokenizer_names:s}"
-    )
```
Comment (sleepinyourhat) [inline jiant/preprocess.py]: Out of curiosity, how could this ever be triggered? If we're allowing people to set task += {tokenizer_name ...}, we should just stop allowing that.

**Comment by W4ngatang** [comment]
Should we just pass around the `Task` object rather than extract attributes? I'm fine either way on this.

File: jiant/preprocess.py [hunk 2]
```diff
@@ -46,13 +46,15 @@
 ALL_SPLITS = ["train", "val", "test"]
 
 
-def _get_serialized_record_path(task_name, split, preproc_dir):
+def _get_serialized_record_path(task_name, tokenizer_name, split, preproc_dir):
     """Get the canonical path for a serialized task split."""
-    serialized_record_path = os.path.join(preproc_dir, "{:s}__{:s}_data".format(task_name, split))
+    serialized_record_path = os.path.join(
+        preproc_dir, "{:s}__{:s}_{:s}_data".format(task_name, tokenizer_name, split)
```
Comment (iftenney) [inline jiant/preprocess.py]: Use double-underscores for the second and third separators as well? I don't think we rigorously enforce that names don't contain underscores elsewhere.

**Comment by iftenney** [comment]
> Should we just pass around the `Task` object rather than extract attributes? I'm fine either way on this.

Passing attributes is safer (pass only what you need).

**Comment by sleepinyourhat** [comment]
(See comment on #866 - If this does enforce one tokenization per task in the directory, then it doesn't solve the problem. I hadn't noticed that that was an expectation.)

**Comment by W4ngatang** [comment]
Does this actually solve the issue? I think it will still break because if we find a cached vocabulary, we load it, even if it doesn't contain a vocabulary relevant to the tokenization we have now, and then when we index with this "vocabulary" (really just PAD and UNK), we get all UNKs.

**Comment by pruksmhc** [comment]
The thinking behind this is that it would be a way for users to be able to structure experiment directories how they want (for example, having on experiment be made up of a task, with various runs representing various preprocessing). Ah, I do see the way vocab is currently handled must also be changed. 
In this case, looks like its a bit more involved to actually change the way we're currently handling experiments and tokenization, so I'll change this to throwing asserts if there are multiple tokenizations being used in an experiment dir.

File: jiant/preprocess.py [hunk 3]
```diff
@@ -396,7 +396,19 @@ def _get_task(name, args, data_path, scratch_path):
...
+
+    task_pkls = glob.glob(os.path.join(scratch_path, "tasks", "%s*" % name))
     pkl_path = os.path.join(scratch_path, "tasks", f"{name:s}.{args.tokenizer:s}.pkl")
+
+    # There should only exist one task pickle per experiment directory with only one
+    # preprocessing, thus check that this is the case if the task pickle directory exists.
+    assert (
+        not os.path.isdir(os.path.join(scratch_path, "tasks"))
+        or len(task_pkls) == 0
+        or args.tokenizer in task_pkls[0]
```
Comment (sleepinyourhat) [inline jiant/preprocess.py]: It seems plausible that for future models, one tokenizer name might be a substring of another, (maybe 'bert-base-cased' and 'multilingual-bert-base-cased'), in which case this check will pass. Mind extracting the exact substring? (Can you do this with pkl_path?)

**Comment by iftenney** [comment]
re: experiment directories: I think it's only an abstraction worth supporting if it allows meaningful sharing of preprocessing, model weights, etc. Otherwise, the user can always just nest them inside their own directory structure, and jiant need not be aware of (nor try to avoid) the other experiments. (This is what I do for edge probing: https://github.com/nyu-mll/jiant/blob/master/scripts/edgeprobing/kubernetes_run_all.sh - there's a `project_dir` with many `exp_dir` inside it, each of which contains exactly one run.

File: jiant/preprocess.py (hunk 3 above)
Comment (iftenney) [reply jiant/preprocess.py]: +1. This kind of bug happens more than you think :)

**Comment by sleepinyourhat** [comment]
Checking in: How close is this to ready?

**Comment by sleepinyourhat** [comment]
Could you either finish or close this this week?

**Comment by sleepinyourhat** [comment]
@pruksmhc - Mind taking a look at this sometime soon?

**Pull request closed without merging**